# Lab 4.03 - Bivariate Analysis of Qualitative Data


In [1]:
# Importing the necessary packages
import numpy as np                                  # "Scientific computing"
import scipy.stats as stats                         # Statistical tests

import pandas as pd                                 # Data Frame
from pandas.api.types import CategoricalDtype

import matplotlib.pyplot as plt                     # Basic visualisation
from statsmodels.graphics.mosaicplot import mosaic  # Mosaic diagram
import seaborn as sns                               # Advanced data visualisation

## Exercise 3 - Discrimination in schoolteacher hiring

African Americans in a St. Louis suburb sued the city 
claiming they were discriminated against in schoolteacher hiring. Of the city's population, 5.7% were 
African American; of 405 teachers in the school system, 15 were African American. Set up appropriate 
hypotheses and determine whether African Americans 
are underrepresented.  
Calculate the standardized residuals. 

Results of the main calculations:
- Chi-squared        χ² = 3.0027
- Critical value      g = 3.8415
- p-value             p = 0.0831
- standardized residuals for african american = -1.7328 > - 2

H0: African Americans are not underrepresented in the school system.  
H1: African Americans are underrepresented in the school system.

In [9]:
types = ['others', 'african american']
obs = np.array([390, 15])
exp_p = np.array([0.943, 0.057])

exp = exp_p * sum(obs)
print(exp)

[381.915  23.085]


In [13]:
alpha = 0.05
n = sum(obs)
k = len(obs)
dof = k - 1
g = stats.chi2.isf(alpha, df=dof)

chi2, p = stats.chisquare(f_obs=obs, f_exp=exp)
sr = (15 - 23) / np.sqrt(23 * (1 - 0.057))

print(round(chi2, 3))
print(round(p, 3))
print(round(g, 3))
print(round(sr, 3))

3.003
0.083
3.841
-1.718


Conclusion: the standarized residual for african americans is -1.718. This means they are less represented than expected, but not significantly. We do not reject H0.